common/oneshot_broadcast/src/lib.rs

```diff
@@ -0,0 +1,182 @@
+//! Provides a single-sender, multiple receiver one-shot channel where any message sent will be
+//! received by all senders.
+//!
+//! This implementation may not be blazingly fast but it should be simple enough to be reliable.
+use parking_lot::{Condvar, Mutex};
+use std::sync::{Arc, Weak};
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Error {
+    SenderDropped,
+}
+
+enum Future<T> {
+    /// The future is ready and the item may be consumed.
+    Ready(T),
+    /// Future is not ready. The contained `Weak` is a reference to the `Sender` that may be used to
+    /// detect when the channel is disconnected.
+    NotReady(Weak<()>),
+}
+
+struct MutexCondvar<T> {
+    mutex: Mutex<Future<T>>,
+    condvar: Condvar,
+}
+
+/// The sending pair of the `oneshot` channel.
+pub struct Sender<T>(Arc<MutexCondvar<T>>, Option<Arc<()>>);
+
+impl<T> Sender<T> {
+    /// Send a message, consuming `self` and delivering the message to *all* receivers.
+    pub fn send(self, item: T) {
+        *self.0.mutex.lock() = Future::Ready(item);
+        // Condvar notification will be handled by the `Drop` implementation.
+    }
+}
+
+impl<T> Drop for Sender<T> {
+    /// Drop the `Arc` and notify all receivers so they can't upgrade their `Weak`s and know that
+    /// the sender has been dropped.
+    fn drop(&mut self) {
+        self.1 = None;
+        self.0.condvar.notify_all();
+    }
+}
+
+/// The receiving pair of the `oneshot` channel. Always receives the message sent by the `Sender`
+/// (if any).
+#[derive(Clone)]
+pub struct Receiver<T: Clone>(Arc<MutexCondvar<T>>);
+
+impl<T: Clone> Receiver<T> {
+    /// Check to see if there is a message to be read *without* blocking/waiting.
+    pub fn try_recv(&self) -> Result<Option<T>, Error> {
+        match &*self.0.mutex.lock() {
```

Comment:
Maybe this should use [`try_lock`](https://docs.rs/lock_api/latest/lock_api/struct.Mutex.html#method.try_lock) and return early if the mutex can't be obtained? That would better satisfy the comment about not blocking.

Or maybe we just clarify the doc comment that this method can block on the mutex, but not to wait for a value

Reply:
Good point! I've clarified the documentation since I think returning early with a busy mutex would be challenging to handle downstream with committee promises.

Reply:
I think the current logic would handle it gracefully by returning the promise from `CommitteeCache::get`, and then calling `wait` on it? But I'm also happy to keep things simple as is